File: c++/BestTimetoBuyandSellStock.cpp
/*
You are given an array prices where prices[i] is the price of a given stock on the ith day.
You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
*/
#include <iostream>
#include <vector>

using namespace std;

class BestTimetoBuyandSellStock {
public:
    int maxProfit(vector<int>& prices) {
    int minPrice = INT_MAX;
    int maxProfit = 0;

    for(int i=0; i<prices.size(); i++){
        if(prices[i] < minPrice){
            minPrice = prices[i];
        }
        else if(prices[i] - minPrice > maxProfit){
            maxProfit = prices[i] - minPrice;
        }
    }

    return maxProfit;
}
};
 

int main() {
    BestTimetoBuyandSellStock b;
    vector <int> myVec = {1, 2, 3, 4, 5}; 
    cout << b.maxProfit(myVec) << endl;
}

File: array/python/BestTimetoBuyandSellStock.py
import sys

class Solution(object):
    def maxProfit(self, prices):
        minPrice = sys.maxsize  
        maxProfit = 0

        for i in range(len(prices)): # [7,1,5,3,6,4] | [7,6,4,3,1]
            if prices[i] < minPrice:
                minPrice = prices[i]
            elif prices[i] - minPrice > maxProfit:
                maxProfit = prices[i] - minPrice
        
        return maxProfit
    
prices = [7,1,5,3,6,4]
solution = Solution()
print(solution.maxProfit(prices)) # Output: 5